python/cudf/cudf/core/index.py

```diff
@@ -307,19 +307,118 @@ def equals(self, other):
+        to the new index.
+
+        Parameters
+        ----------
+        other : Index.
+        how : {'left', 'right', 'inner', 'outer'}
+        return_indexers : bool, default False
+        sort : bool, default False
+            Sort the join keys lexicographically in the result Index. If False,
+            the order of the join keys depends on the join type (how keyword).
+
+        Returns: index
+
+        Examples
+        --------
+        >>> import cudf
+        >>> lhs = cudf.DataFrame(
+        ...     {"a":[2, 3, 1], "b":[3, 4, 2]}).set_index(['a', 'b']
+        ... ).index
+        >>> rhs = cudf.DataFrame({"a":[1, 4, 3]}).set_index('a').index
+        >>> lhs.join(rhs, how='inner')
+        MultiIndex(levels=[0    1
+        1    3
+        dtype: int64, 0    2
+        1    4
+        dtype: int64],
+        codes=   a  b
+        0  1  1
+        1  0  0)
+        """
+
+        if isinstance(self, cudf.MultiIndex) and isinstance(
+            other, cudf.MultiIndex
+        ):
+            raise TypeError(
+                "Join on level between two MultiIndex objects is ambiguous"
+            )
+
+        if level is not None and not is_scalar(level):
+            raise ValueError("level should be an int or a label only")
+
+        # `_merge` doesn't support `right`
+        if how == "right":
+            how = "left"
+            lhs = other.copy(deep=False)
+            rhs = self.copy(deep=False)
+        else:
+            lhs = self.copy(deep=False)
+            rhs = other.copy(deep=False)
+
+        on = level
+        left_name = lhs.name
+        right_name = rhs.name
+        # There should be no None values in Joined indices,
+        # so essentially it would be `left` or 'inner'
+        if isinstance(lhs, cudf.MultiIndex):
+            if level is not None and isinstance(level, int):
+                on = lhs._data.get_by_index(level).names[0]
+            right_name = on or right_name
+            on = right_name
+            how = "left" if how == "outer" else how
+
+        elif isinstance(rhs, cudf.MultiIndex):
+            if level is not None and isinstance(level, int):
+                on = rhs._data.get_by_index(level).names[0]
+            left_name = on or left_name
+            on = left_name
+            if how == "left":
+                how = "inner"
+            elif how == "outer":
+                how = "left"
+
+            lhs, rhs = rhs, lhs
+            left_name, right_name = right_name, left_name
         else:
-            return column_join_res
+            # Both are nomal indices
+            right_name = left_name
+            on = right_name
+
+        def _set_categories(col, cats):
+            return col.cat()._set_categories(cats, is_unique=True).fillna(-1)
+
+        lhs = lhs.to_frame(index=False, name=left_name)
+        rhs = rhs.to_frame(index=False, name=right_name)
+
+        # Handle categorical columns
```

Comment:
What happens if you remove this code? `Frame._merge` *should* already be handling the categorical conversions here.

Reply:
`Frame._merge` expects both columns to have same set of categories.

Reply:
Can you try merging the tip of 0.15? Join refactor was merged yesterday - it should allow for mixed categoricals. Apologies if this breaks all your hard work here.

Reply:
Oh, ok. I didn't try that. I will take a look.